### ISSUE TITLE
add support for ac power supply

### ISSUE THREAD

**Issue by ael-code** [i3/i3status#180]
It would be really useful for me to see if the AC power supply is plugged and working. What do you think about adding support for it?

My `/sys/class/power_supply/AC/uevent`:
```
POWER_SUPPLY_NAME=AC
POWER_SUPPLY_ONLINE=1
```

**Comment by stapelberg** [comment]
Can you clarify how this differs from your battery not discharging? I.e., the battery is not discharging if and only if the power supply is plugged in, so the two are equivalent, right?

**Comment by ael-code** [comment]
On 11/17/2016 09:00 AM, Michael Stapelberg wrote:
> Can you clarify how this differs from your battery not discharging?
> I.e., the battery is not discharging if and only if the power supply is
> plugged in, so the two are equivalent, right?
> 

Actually is not exactly like that.

Having a look at the [kernel
code](https://github.com/torvalds/linux/blob/5924bbecd0267d87c24110cbe2041b5075173a25/include/linux/power_supply.h)
we could have 5 possible status:

```
POWER_SUPPLY_STATUS_UNKNOWN
POWER_SUPPLY_STATUS_CHARGING,
POWER_SUPPLY_STATUS_DISCHARGING
POWER_SUPPLY_STATUS_NOT_CHARGING
POWER_SUPPLY_STATUS_FULL
```

There are some thinkpad models that support charging thresholds. e.g.
you can set that your battery should not go above the 60% of capacity to
preserve it.

I think that the `POWER_SUPPLY_STATUS_NOT_CHARGING` is exactly to
express this condition. e.g. the battery is at 60%, the external power
supply is plugged in and the battery is not CHARGING.

I think that all of these 5 states should be supported by i3status.

**Comment by stapelberg** [comment]
Let me rephrase: If you want to know whether your power supply is plugged in, check whether i3status displays “Discharging” as status. If not, then your power supply is plugged in.

I don’t think it’s necessary to implement a special case for the not_charging state.

**Comment by ael-code** [comment]
On 11/19/2016 02:54 PM, Michael Stapelberg wrote:
> Let me rephrase: If you want to know whether your power supply is
> plugged in, check whether i3status displays “Discharging” as status. If
> not, then your power supply is plugged in.
> 

That's a good point, but what about the custom status string? I like
them because you can use icons in them that requires less time to read.


Looking at the code I find out that:

The kernel is providing 5 different states, the internal code of
i3status is mapping them in 4 different states and these states are then
mapped to 3 custom messages:

 STATUS_UNKNOWN         -> CS_UNKNOWN     -> status_bat
 STATUS_CHARGING        -> CS_CHARGING    -> status_chr
 STATUS_DISCHARGING     -> CS_DISCHARGING -> status_bat
 STATUS_NOT_DISCHARGING -> CS_UNKNOWN     -> status_bat
 STATUS_FULL            -> CS_FULL        -> status_full

Thus I think that it is impossible, using the custom strings, to
differentiate the two kernel status DISCHARGING and NOT_DISCHARGING.

Am I wrong?

> I don’t think it’s necessary to implement a special case for the
> not_charging state.
>

**Comment by stapelberg** [comment]
> Thus I think that it is impossible, using the custom strings, to
> differentiate the two kernel status DISCHARGING and NOT_DISCHARGING

Correct, you cannot use the `status_*` options by themselves. You would need to use the `%status` format string. Issue #119 is related.
